secrets/azurekeyvault/akv.go

```diff
@@ -0,0 +1,187 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Package azurekeyvault provides a secrets implementation backed by Azure KeyVault.
+// See https://docs.microsoft.com/en-us/azure/key-vault/key-vault-whatis for more information.
+// Use NewKeeper to construct a *secrets.Keeper.
+//
+// As
+//
+// azurekeyvault exposes the following type for As:
+//  - Error: autorest.DetailedError, see https://godoc.org/github.com/Azure/go-autorest/autorest#DetailedError
+package azurekeyvault
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/internal/gcerr"
+	"gocloud.dev/internal/useragent"
+	"gocloud.dev/secrets"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+var (
+	// KeyVault URI suffix
+	keyVaultEndpointSuffix = "vault.azure.net"
+	// Map of HTTP Status Code to go-cloud ErrorCode
+	errorCodeMap = map[int]gcerrors.ErrorCode{
+		200: gcerrors.OK,
+		400: gcerrors.InvalidArgument,
+		401: gcerrors.PermissionDenied,
+		404: gcerrors.NotFound,
+		408: gcerrors.DeadlineExceeded,
+		429: gcerrors.ResourceExhausted,
+		500: gcerrors.Internal,
+		501: gcerrors.Unimplemented,
+	}
+)
+
+type (
+	keeper struct {
+		client       *keyvault.BaseClient
+		keyVaultName string
+		keyName      string
+		keyVersion   string
+		options      *KeeperOptions
+	}
+
+	// KeeperOptions provides configuration options for encryption/decryption operations.
+	KeeperOptions struct {
+		Algorithm string
+	}
+)
+
+// Dial gets a new *keyvault.BaseClient, see https://godoc.org/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#BaseClient
+func Dial() (*keyvault.BaseClient, error) {
+	auth, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	client := keyvault.NewWithoutDefaults()
+	client.Authorizer = auth
+	client.Sender = autorest.NewClientWithUserAgent(useragent.AzureUserAgentPrefix("secrets"))
+	return &client, nil
+}
+
+// NewKeeper returns a *secrets.Keeper that uses Azure keyVault.
+// List of Parameters:
+//	- client: *keyvault.BaseClient instance, see https://godoc.org/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#BaseClient
+//  - keyVaultName: string representing the KeyVault name, see https://docs.microsoft.com/en-us/azure/key-vault/common-parameters-and-headers
+//  - keyName: string representing the keyName, see https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#uri-parameters
+//  - keyVersion: string representing the keyVersion, see https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#uri-parameters
```

Comment:
The link says `keyVersion` is required, is it? In the test there is only one `keyVersion` for both keys and not filled in.

Reply:
It is required in the http request, however the value can be blank which means it will default the the Current Key Version, see this [link ](https://docs.microsoft.com/en-us/azure/key-vault/about-keys-secrets-and-certificates)for more info